File: eval_harness/evaluator/skill_invocation/build_health/bundle_check.py
"""Bundle half of the build-health cascade: does the whole app survive a
real Metro bundle (`expo export`)? Needs the authored app's own
node_modules, which the analysis-time artifact doesn't ship (excluded
before upload) -- so unlike syntax_check, this can NOT run at analysis
time. It runs once, at AUTHORING time (author-app.sh, right after the
coding agent stops, while node_modules still exists), and persists its
result as a small JSON file inside the workspace itself -- which DOES
survive into the artifact -- for analysis to read later.

CLI: python3 -m eval_harness.evaluator.skill_invocation.build_health.bundle_check <workspace-dir>
"""

from __future__ import annotations

import json
from pathlib import Path
import shutil
import subprocess
import sys


RESULT_FILENAME = ".eval-build-health-bundle.json"
_TIMEOUT_SECONDS = 180


def compute_bundle_result(app_dir: Path) -> dict:
    """Never raises -- a failure to even invoke `expo export` (missing
    node_modules, timeout, etc.) is itself a result, not an exception."""
    expo_bin = app_dir / "node_modules" / ".bin" / "expo"
    if not expo_bin.exists():
        return {"ok": None, "reason": "no node_modules/.bin/expo in workspace"}

    export_dir = app_dir / ".eval-bundle-export-tmp"
    try:
        result = subprocess.run(
            [str(expo_bin), "export", "--platform", "ios", "--output-dir", str(export_dir)],
            cwd=app_dir,
            capture_output=True,
            text=True,
            timeout=_TIMEOUT_SECONDS,
        )
    except subprocess.TimeoutExpired:
        return {"ok": False, "reason": f"expo export timed out after {_TIMEOUT_SECONDS}s"}
    except Exception as exc:
        return {"ok": False, "reason": f"failed to invoke expo export: {exc}"}
    finally:
        shutil.rmtree(export_dir, ignore_errors=True)

    if result.returncode != 0:
        tail = (result.stderr or result.stdout or "")[-1500:]
        return {"ok": False, "reason": tail}
    return {"ok": True}


def persist_bundle_result(app_dir: Path) -> dict:
    result = compute_bundle_result(app_dir)
    (app_dir / RESULT_FILENAME).write_text(json.dumps(result))
    return result


def read_bundle_result(app_dir: Path) -> dict | None:
    """Analysis-time read of the result persisted at authoring time. None
    means the authoring-time stage never ran (e.g. an artifact from before
    this feature existed) -- distinct from a real {"ok": False, ...}."""
    path = app_dir / RESULT_FILENAME
    if not path.exists():
        return None
    try:
        return json.loads(path.read_text())
    except json.JSONDecodeError:
        return None


if __name__ == "__main__":
    workspace = Path(sys.argv[1])
    outcome = persist_bundle_result(workspace)
    print(f"bundle check: {outcome}")

File: packages/skill-analyzer/src/build_health/bundle_check.ts
import { existsSync, readFileSync, rmSync, writeFileSync } from 'node:fs';
import { join } from 'node:path';

export const RESULT_FILENAME = '.eval-build-health-bundle.json';
const TIMEOUT_MILLISECONDS = 180_000;

export type BundleResult = { ok: boolean | null; reason?: string };
export type BundlePlatform = 'ios' | 'android';
export type BundleOptions = { platform?: BundlePlatform };

export function computeBundleResult(appDir: string, options: BundleOptions = {}): BundleResult {
  const platform = options.platform ?? 'ios';
  const expoBin = join(appDir, 'node_modules', '.bin', 'expo');
  if (!existsSync(expoBin)) {
    return { ok: null, reason: 'no node_modules/.bin/expo in workspace' };
  }
  const exportDir = join(appDir, '.eval-bundle-export-tmp');
  try {
    const result = Bun.spawnSync(
      [expoBin, 'export', '--platform', platform, '--output-dir', exportDir],
      {
        cwd: appDir,
        stdout: 'pipe',
        stderr: 'pipe',
        timeout: TIMEOUT_MILLISECONDS,
      }
    );
    if (result.exitedDueToTimeout === true) {
      return {
        ok: false,
        reason: `expo export timed out after ${TIMEOUT_MILLISECONDS / 1000}s`,
      };
    }
    if (result.exitCode !== 0) {
      const stderr = new TextDecoder().decode(result.stderr);
      const stdout = new TextDecoder().decode(result.stdout);
      return { ok: false, reason: (stderr || stdout).slice(-1500) };
    }
    return { ok: true };
  } catch (error) {
    const detail = error instanceof Error ? error.message : String(error);
    return { ok: false, reason: `failed to invoke expo export: ${detail}` };
  } finally {
    try {
      rmSync(exportDir, { recursive: true, force: true });
    } catch {
      // Match shutil.rmtree(..., ignore_errors=True): cleanup must never hide
      // the build-health result we already computed.
    }
  }
}

export function persistBundleResult(appDir: string, options: BundleOptions = {}): BundleResult {
  const result = computeBundleResult(appDir, options);
  writeFileSync(join(appDir, RESULT_FILENAME), JSON.stringify(result), 'utf8');
  return result;
}

export function readBundleResult(appDir: string): BundleResult | null {
  const path = join(appDir, RESULT_FILENAME);
  if (!existsSync(path)) return null;
  try {
    return JSON.parse(readFileSync(path, 'utf8')) as BundleResult;
  } catch {
    return null;
  }
}

export function main(args: string[] = process.argv.slice(2)): number {
  const [workspace, ...options] = args;
  if (workspace === undefined) {
    console.error('usage: skill-analyzer bundle-check <workspace-dir> [--platform ios|android]');
    return 2;
  }
  let platform: BundlePlatform = 'ios';
  if (options.length === 2 && options[0] === '--platform') {
    const requestedPlatform = options[1];
    if (requestedPlatform !== 'ios' && requestedPlatform !== 'android') {
      console.error('error: --platform must be ios or android');
      return 2;
    }
    platform = requestedPlatform;
  } else if (options.length !== 0) {
    console.error('usage: skill-analyzer bundle-check <workspace-dir> [--platform ios|android]');
    return 2;
  }
  const outcome = persistBundleResult(workspace, { platform });
  console.log(`bundle check: ${JSON.stringify(outcome)}`);
  return 0;
}

if (import.meta.main) {
  process.exitCode = main();
}
